### PR TITLE
Customers SAGA

### PR DESCRIPTION
**Pull request by andiserg**

### CODE REVIEW

**Comment by andiserg** [comment]
image

This is an implementation of choreographed SAGA that helps to handle distributed transactions. There are two operations in this project where saga is used: client creation and deletion endpoints. The customer module starts processing the request, creates and stores the client data in the database, then generates an event and waits until it receives a SubscriberCreated or SubscriberCreatedError from the notification service, and according to the event, returns a 200 or 500 response to the user.

Is it really a choreographed saga? Yes, it is, because the endpoints of the customer module do not manage the entire transaction, but wait for the final message to return a response to the user. All events are processed asynchronously.

**Comment by andiserg** [comment]
image

In the event of an error, the customer consumer receives an error event and takes appropriate actions like compensation transaction. The endpoint handler also receives an event and returns bad response.

How does the consumer know which transaction needs to be compensated? All events have a `transactionID` field that is generated at the entry point (in this case, in the client handler). The pull consumer in the handler takes the latest events from the `events.<transaction_id>` queue and creates a response according to the result.

File: internal/notifier/app/consumer.go [hunk 1]
```diff
@@ -1,56 +1,81 @@
...
+
+func newMessageHandler(container InteractorFactory) func(msg jetstream.Msg) {
+	return func(msg jetstream.Msg) {
+		log.Println("Consumed event")
 		var event Message
-		err := json.Unmarshal(msg.Data, &event)
+		err := json.Unmarshal(msg.Data(), &event)
 		if err != nil {
 			log.Printf("Error unmarshalling message: %v", err)
+			_ = msg.Nak()
```
Comment (andriihar) [inline internal/notifier/app/consumer.go:58]: we really need it here?

File: internal/notifier/app/consumer.go [hunk 2]
```diff
@@ -1,56 +1,81 @@
...
 			return
 		}
 
 		switch event.Title {
 		case "UserCreated":
-			userCreatedHandle(event, c.container)
+			interactor := container.CreateSubscriber()
+			inputData := createsubscriber.InputData{
+				Email:         event.Data["email"].(string),
```
Comment (andriihar) [inline internal/notifier/app/consumer.go:66]: Can we use struct for data?

File: internal/rateservice/customers/app/consumer.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	js        jetstream.JetStream
+	container presentation.InteractorFactory
+}
+
+func NewConsumer(ctx context.Context, js jetstream.JetStream, container *IoC) Consumer {
+	return Consumer{ctx: ctx, js: js, container: container}
+}
+
+func (c Consumer) Run() jetstream.ConsumeContext {
+	_, _ = broker.NewStream(c.ctx, c.js, "events")
```
Comment (andriihar) [inline internal/rateservice/customers/app/consumer.go:33]: from my perspective, better to put context to method, not to struct.
https://go.dev/blog/context-and-structs#prefer-contexts-passed-as-arguments

**Pull request merged**